### PR TITLE
[Perf] Add regional compilation support for SenseNova-U1

### PR DESCRIPTION
**Pull request by akshatvishu**
## Purpose
Enable regional `torch.compile` support for SenseNova-U1 by declaring `SenseNovaU1DecoderLayer` in` _repeated_blocks` , pre-allocating `DynamicCache` layers before entering compiled decoder blocks and passing fixed cache layers into each block to avoid mutating the cache container during tracing.

## Test Plan
```py
# SPDX-License-Identifier: Apache-2.0
"""Measure SenseNova-U1 cold and steady-state generation latency."""

from __future__ import annotations

import argparse
import statistics
import time
from pathlib import Path

import torch

from vllm_omni.entrypoints.omni import Omni
from vllm_omni.inputs.data import OmniDiffusionSamplingParams

DEFAULT_PROMPT = (
    "Close portrait of an elderly woman by a farmhouse window, textured skin, gentle smile, warm natural light, "
    "emotional documentary look. The portrait should feel polished and natural, with sharp eyes, realistic skin "
    "texture, accurate facial anatomy, and premium lighting that keeps the face as the main focus."
)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--model", default="SenseNova/SenseNova-U1-8B-MoT")
    parser.add_argument("--prompt", default=DEFAULT_PROMPT)
    parser.add_argument("--width", type=int, default=1536)
    parser.add_argument("--height", type=int, default=2720)
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--num-steps", type=int, default=50)
    parser.add_argument("--cfg-scale", type=float, default=4.0)
    parser.add_argument("--timestep-shift", type=float, default=3.0)
    parser.add_argument("--cfg-norm", default="none")
    parser.add_argument("--warmup-runs", type=int, default=1)
    parser.add_argument("--measured-runs", type=int, default=3)
    parser.add_argument("--output-dir", type=Path, default=Path("outputs"))
    parser.add_argument("--enforce-eager", action="store_true")
    parser.add_argument("--tensor-parallel-size", type=int, default=1)
    parser.add_argument("--think", action=argparse.BooleanOptionalAction, default=True)
    parser.add_argument("--print-think", action=argparse.BooleanOptionalAction, default=True)
    return parser.parse_args()


def synchronize() -> None:
    if torch.accelerator.is_available():
        torch.accelerator.synchronize()


def generate_once(
    omni: Omni,
    prompt: dict[str, object],
    sampling_params: OmniDiffusionSamplingParams,
) -> tuple[list[object], float]:
    synchronize()
    start = time.perf_counter()
    outputs = list(omni.generate(prompts=prompt, sampling_params_list=sampling_params))
    synchronize()
    return outputs, time.perf_counter() - start


def main() -> None:
    args = parse_args()
    if args.warmup_runs < 0:
        raise ValueError("--warmup-runs must be non-negative")
    if args.measured_runs < 1:
        raise ValueError("--measured-runs must be positive")

    mode = "eager" if args.enforce_eager else "regional-compile"
    print(f"Mode: {mode}")

    omni = Omni(
        model=args.model,
        tensor_parallel_size=args.tensor_parallel_size,
        enforce_eager=args.enforce_eager,
    )
    prompt = {"prompt": args.prompt, "modalities": ["image"]}
    sampling_params = OmniDiffusionSamplingParams(
        height=args.height,
        width=args.width,
        seed=args.seed,
        num_inference_steps=args.num_steps,
        extra_args={
            "cfg_scale": args.cfg_scale,
            "cfg_norm": args.cfg_norm,
            "timestep_shift": args.timestep_shift,
            "cfg_interval": (0.0, 1.0),
            "batch_size": 1,
            "think": args.think,
            "t_eps": 0.02,
        },
    )

    outputs, cold_latency = generate_once(omni, prompt, sampling_params)
    print(f"Cold latency: {cold_latency:.3f} s")

    for run in range(args.warmup_runs):
        _, latency = generate_once(omni, prompt, sampling_params)
        print(f"Warmup {run + 1}: {latency:.3f} s")

    latencies = []
    for run in range(args.measured_runs):
        outputs, latency = generate_once(omni, prompt, sampling_params)
        latencies.append(latency)
        print(f"Measured {run + 1}: {latency:.3f} s")

    mean = statistics.fmean(latencies)
    stddev = statistics.stdev(latencies) if len(latencies) > 1 else 0.0
    print(f"Steady-state latency: {mean:.3f} ± {stddev:.3f} s")

    args.output_dir.mkdir(parents=True, exist_ok=True)
    for output in outputs:
        custom = getattr(output, "_custom_output", {}) or {}
        if args.print_think and custom.get("think_text"):
            print(f"[Think]\n{custom['think_text']}\n")
        for index, image in enumerate(getattr(output, "images", None) or []):
            path = args.output_dir / f"sensenova_{mode}_{index}.png"
            image.save(path)
            print(f"Saved {path}")


if __name__ == "__main__":
    main()

```
## ENV:
Official Docker ROcm Image  + single MI300x 

ROCM Version                 : 7.2.53211-c2d9476115
vLLM Version                 : 0.23.0
vLLM-Omni Version            : 0.1.dev2062+gdf159b88d.rocm (git sha: df159b88d, date: ocm)

## Test Result

```bash
INFO 06-25 21:01:43 [compile.py:64] Regional compilation applied to 42 module(s) for repeated blocks ['SenseNovaU1DecoderLayer'].
INFO 06-25 21:01:43 [diffusion_model_runner.py:93] Model runner: transformer compiled with torch.compile.
```

#### Environment / Workload

| Item | Value |
|---|---|
| Model | `SenseNova/SenseNova-U1-8B-MoT` |
| Platform | ROCm detected |
| Batch size | 1 |
| Output type | image |
| Image grid | `[[32, 32]]` |
| Compile mode | regional compile over 42 `SenseNovaU1DecoderLayer` blocks |
| Eager flag | `--enforce-eager` |

---

## Summary

| Metric | Eager | Compile | Delta |
|---|---:|---:|---:|
| Engine init time | 27.02 s | 69.71 s | Compile +42.69 s slower init |
| Model load memory | 32.7740 GiB | 32.7740 GiB | Same |
| Model load time | 13.337203 s | 12.569961 s | Compile 0.767 s faster |
| Cold latency | 40.203 s | 45.989 s | Compile 5.786 s slower |
| Warmup 1 latency | 37.576 s | 25.798 s | Compile 11.778 s faster |
| Measured 1 | 37.444 s | 25.514 s | Compile 11.930 s faster |
| Measured 2 | 37.455 s | 25.484 s | Compile 11.971 s faster |
| Measured 3 | 37.332 s | 25.545 s | Compile 11.787 s faster |
| Steady-state latency | 37.411 ± 0.068 s | 25.514 ± 0.031 s | Compile 11.897 s faster |
| Steady-state speedup | 1.00x | 1.47x | +46.6% throughput |
| Steady-state latency reduction | baseline | 31.8% lower | -31.8% |
| Peak reserved GPU memory | 35.79 GB | 35.58 GB | Compile 0.21 GB lower |
| Peak allocated GPU memory | 35.13 GB | 35.03 GB | Compile 0.10 GB lower |
| Pool overhead | 0.65 GB / 1.8% | 0.56 GB / 1.6% | Compile slightly lower |

---

#### Interpretation

Compile mode has a much higher startup cost:

```text
Eager engine init:   27.02 s
Compile engine init: 69.71 s
Delta:              +42.69 s
```

Compile mode also has a slower first real request:

```text
Eager cold latency:   40.203 s
Compile cold latency: 45.989 s
Delta:                +5.786 s
```
However, after warmup, compile mode is substantially faster:

```text
Eager steady-state:   37.411 ± 0.068 s
Compile steady-state: 25.514 ± 0.031 s
Speedup:              1.47x
Latency reduction:    31.8%
```
Memory usage is effectively similar, with compile mode slightly lower in the measured steady-state requests:

```text
Eager:   35.79 GB reserved / 35.13 GB allocated
Compile: 35.58 GB reserved / 35.03 GB allocated
```
#### Eager 
sensenova_eager_0

#### Compile 
sensenova_regional-compile_0

---
<a href="https://app.blacksmith.sh/vllm-project/codesmith/vllm-omni/pr/4732"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1785015910&installation_id=142609222&pr_number=4732&repository=vllm-project%2Fvllm-omni&return_to=https%3A%2F%2Fgithub.com%2Fvllm-project%2Fvllm-omni%2Fpull%2F4732&signature=dbda13a15501cc6e000368238ce926ef8fbb7c53df4959274df2186e8b344fd0"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

**Comment by akshatvishu** [comment]
#### Regression Testing
```bash
pytest -q \
    tests/e2e/offline_inference/test_sensenova_u1_text2img.py \
    tests/e2e/offline_inference/test_sensenova_u1_img2img.py \
    tests/e2e/offline_inference/test_sensenova_u1_expansion.py
```

image

```bash
pytest -q -m cpu \
    tests/diffusion/models/sensenova_u1/test_sensenova_u1_compile.py \
    tests/model_extras/test_model_extras.py -k sensenova
```

image
